Update Subdomain gateway section (#431)
ipfs/ipfs-docs: docs/how-to/address-ipfs-on-web.md

### Subdomain gateway

> When [origin-based security](https://en.wikipedia.org/wiki/Same-origin_policy) perimeter is needed, [CIDv1](/concepts/content-addressing/#identifier-formats) in case-insensitive encoding such as Base32 or Base36 should be used in the subdomain:

Author: I've used "security perimeter" as in https://en.wikipedia.org/wiki/Perimeter_security, but if it sounds weird, we can drop "perimeter" and just have:

```suggestion
When [origin-based security](https://en.wikipedia.org/wiki/Same-origin_policy) is needed, [CIDv1](/concepts/content-addressing/#identifier-formats) in case-insensitive encoding such as Base32 or Base36 should be used in the subdomain:
```

Thoughts @johnnymatthews?

    https://<cidv1b32>.ipfs.<gateway-host>.tld/path/to/resource

Example:

    https://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq.ipfs.dweb.link/wiki/
    https://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq.ipfs.cf-ipfs.com/wiki/Vincent_van_Gogh.html
    https://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq.ipfs.localhost:8080/wiki/

> ::: tip Native support in go-ipfs 0.5+

Reviewer: ```suggestion
#### Native support in go-ipfs 0.5+
```
The sections marked out with `:::` turn into colored boxes in the rendered VuePress site, so we should use them sparingly.

> [go-ipfs](https://dist.ipfs.io/#go-ipfs) provides native support for subdomain gateways on hostnames defined in [`Gateway.PublicGateways`](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#gatewaypublicgateways) configuration map.

Reviewer: ```suggestion
[go-ipfs](https://dist.ipfs.io/#go-ipfs) provides native support for subdomain gateways on hostnames defined in the [`Gateway.PublicGateways`](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#gatewaypublicgateways) configuration map.
```

Learn more about daemon configuration for hosting a public gateway:

- [`Gateway.PublicGateways` docs](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#gatewaypublicgateways) for defining gateway behavior on specified hostnames
- [`Gateway` recipes](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#gateway-recipes) with ready to use one-liners for most common use cases

> :::

Reviewer: ```suggestion
```

::: warning Known issues

> - Some browsers and other user agents force lowercase for the authority part of URLs, breaking case-sensitive CIDs before HTTP Gateway has a chance to read them.

Reviewer: ```suggestion
- Some browsers and other user agents force lowercase for the authority part of URLs, breaking case-sensitive CIDs before the HTTP gateway has a chance to read them
```

- DNS label length is limited to 63 characters ([RFC 1034](https://tools.ietf.org/html/rfc1034#page-7))

> Due to these limitaitons, use of short, case-insensitive CIDv1 in subdomain context is advised.

Reviewer 2: Are we advising people that they should canonically tell people to use `gateway.tld/path` and that it will redirect to the subdomain or is this unrelated?

Author: Yes. I've included that in _CID conversion for subdomains_ section below. 
Should we make this more prominent somehow?

Reviewer 2: The question is more about what we're telling people. Do we tell them "hey, CIDs in subdomains is a security feature but the correct way for people to talk about URLs is using paths", or say "there are two valid ways to describe URLs"?

If it's the former then we should be louder/more prominent and explain that we'd like people to use CIDv1s so that things look nicer bc we redirect localhost:8080/ipfs/{cid} to {cid}.ipfs.localhost instead of to {cid in different representation}.ipfs.localhost. If it's the latter then we don't have to do anything.

What's your feeling on this?

Author: @aschmahmann I think paths make it easier  to grasp the concept of "content addressing", so my intuition is that we should be thinking in therms of  "hey, CIDs in subdomains is a security feature but the correct way for people to talk about URLs is using paths". So `gateway.tld/path` remains "canonical". 

Not sure if we need to be more vocal: I believe people who care about subdomains already default to CIDv1, so the hash remains the same. What we cna do, is to ensure our GUIs copy path version instead of subdomain version etc.

Reviewer 2: I think your approach is reasonable here as long as we're also trying to ensure that our comms focus on path domains. Sometimes subdomains will inevitably slip in as someone copy-pastes `cid.ipfs.localhost` from their URL bar and posts it somewhere, but not much to be done there at the moment.

> Due to these limitaitons, use of short, case-insensitive CIDv1 in subdomain context is advised.

Reviewer: ```suggestion
Due to these limitations, the use of short, case-insensitive CIDv1 in a subdomain context is advised.
```

> Base32 is the safe default. Less popular Base36 can be used for longer ED25519 libp2p keys.

Reviewer 2: I don't know what our end game is going to be around base representation for content CID (e.g. we may want to default to b36 everywhere for consistency), but this seems reasonable for now

> Base32 is the safe default. Less popular Base36 can be used for longer ED25519 libp2p keys.

Reviewer: ```suggestion
Base32 is the safe default; the less-popular Base36 can be used for longer ED25519 libp2p keys.
```

> See the next section to learn how to convert existing CID v0 to a DNS-safe representation.

Reviewer: ```suggestion
See the next section to learn how to convert an existing CIDv0 to a DNS-safe representation.
```

:::

> ::: tip CID conversion for subdomains

Reviewer: ```suggestion
#### CID conversion for subdomains
```

> If you have content identified by an older CID v0, there are easy ways to safely represent it as CID v1 for use in subdomains and other case-insensitive contexts.

Reviewer: ```suggestion
If you have content identified by an older CIDv0, there are easy ways to safely represent it as CIDv1 for use in subdomains and other case-insensitive contexts.
```

> #### Automatic – leverage the gateway in go-ipfs

Reviewer: ```suggestion
#### Automatic — leverage the gateway in go-ipfs
```

> **TL;DR:** using subdomain gateway as a drop-in replacement for a path one removes the need for manual CID conversion.

Reviewer: ```suggestion
**TL;DR:** Using a subdomain gateway as a drop-in replacement for a path one removes the need for manual CID conversion.
```

> Request for a content path sent to the gateway domain will return HTTP 301 redirect to a correct subdomain version, taking care of any necessary encoding conversion, if needed:

Reviewer: ```suggestion
Request for a content path sent to the gateway domain will return an HTTP 301 redirect to a correct subdomain version, taking care of any necessary encoding conversion if needed:
```

    https://<gateway-host>.tld/ipfs/<cid> -> https://<cidv1>.ipfs.<gateway-host>.tld/

> To illustrate, opening the CID v0 resource at [`https://dweb.link/ipfs/QmT5NvUtoM5nWFfrQdVrFtvGfKFmG7AHE8P34isapyhCxX/wiki/Mars.html`](https://dweb.link/ipfs/QmT5NvUtoM5nWFfrQdVrFtvGfKFmG7AHE8P34isapyhCxX/wiki/Mars.html)

Reviewer: ```suggestion
To illustrate, opening the CIDv0 resource at [`https://dweb.link/ipfs/QmT5NvUtoM5nWFfrQdVrFtvGfKFmG7AHE8P34isapyhCxX/wiki/Mars.html`](https://dweb.link/ipfs/QmT5NvUtoM5nWFfrQdVrFtvGfKFmG7AHE8P34isapyhCxX/wiki/Mars.html)
```

> returns a redirect to a CID v1 representation at [`https://bafybeicgmdpvw4duutrmdxl4a7gc52sxyuk7nz5gby77afwdteh3jc5bqa.ipfs.dweb.link/wiki/Mars.html`](https://bafybeicgmdpvw4duutrmdxl4a7gc52sxyuk7nz5gby77afwdteh3jc5bqa.ipfs.dweb.link/wiki/Mars.html).

Reviewer: ```suggestion
returns a redirect to a CIDv1 representation at [`https://bafybeicgmdpvw4duutrmdxl4a7gc52sxyuk7nz5gby77afwdteh3jc5bqa.ipfs.dweb.link/wiki/Mars.html`](https://bafybeicgmdpvw4duutrmdxl4a7gc52sxyuk7nz5gby77afwdteh3jc5bqa.ipfs.dweb.link/wiki/Mars.html).
```

> Gateway takes care of converting CID to case-insensitive encoding.

Reviewer: ```suggestion
The gateway takes care of converting the CID to case-insensitive encoding.
```

> The multihash in CID v1 is the same as in original CID v0.

Reviewer: ```suggestion
The multihash in CIDv1 is the same as in the original CIDv0.
```

> #### Manual – use CID website or commandline

Reviewer: ```suggestion
#### Manual — use cid.ipfs.io or the command line
```

One can also do the conversion manually.

> To convert CID to Base32 ([RFC4648](https://tools.ietf.org/html/rfc4648#section-6), no padding) use [cid.ipfs.io](https://cid.ipfs.io) or a command line:

Reviewer: ```suggestion
To convert a CID to Base32 ([RFC4648](https://tools.ietf.org/html/rfc4648#section-6), no padding) use [cid.ipfs.io](https://cid.ipfs.io) or the command line:
```

```shell
> ipfs cid base32 QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR
bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi
```

PeerIDs can be represented as [CID with `libp2p-key` multicodec](https://github.com/libp2p/specs/blob/master/RFC/0001-text-peerid-cid.md).
Base36 is suggested as a safer default for longer keys:

```shell
> ipfs cid format -v 1 -b base36 --codec libp2p-key QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN

> k2k4r8jl0yz8qjgqbmc2cdu5hkqek5rj6flgnlkyywynci20j0iuyfuj

Reviewer 2: This only works on RSA keys, Ed25519 keys fail here. Ed25519 keys aren't super widely used right now, but as go-ipfs is switching to using them by default and PeerIDs will still be encoded as base58 by default we might want to include.

`ipfs id --format "<id>" --peerid-base base36` or using `ipfs key list -l` (we unfortunately haven't added support yet for `ipfs key list -l mykey`, but hopefully that'll make it into either the final release or 0.8.0)

Author: Eek. This edge case again :( 

Documenting this won't help much in practice: people will have pretty bad UX when `ipfs cid format` fails on ED25519 without an obvious reason, only a subset will read the docs and understand whats happening.

AFAIK `ipfs id` is the only place with non-CID PeerID when Ed25519 is used. 
`ipfs key list -l` prints Base36 version for `self` already.

@aschmahmann : how do you feel about `ipfs id` returning CIDv1b36 when PeerID is an identity multihash? It will close the UX gap.

I know its special-casing, but imo its worth it: will save us a lot of headache and time on user support in the following months.

Reviewer 2: The issue is that the libp2p ecosystem is unprepared for peerIDs as CIDs and aside from this most recent go-libp2p (and go-ipfs) release nobody can process multiaddresses like `/p2p/k51abc...` which means the "normal" user flow of `ipfs swarm connect {copy-paste ipv4 part of multiaddress}/p2p/{copy-paste peerID} won't work on older nodes.

How commonly/why would people be converting the representations of peerIDs anyway? In my understanding using peerIDs in subdomains is not yet a commonly used feature. As a result my inclination is to just let this edge case be a bit miserable for a while and fix it later.

Some discussion on options around dealing with this are https://github.com/libp2p/go-libp2p/issues/976.

@Stebalien WDYT?

Author: Ack, I think p2p tunel is still experimental, so the only real life use will be IPNS on subdomains.
We already encourage people to publish to IPNS using a key other than `self`, so you are right, should be fine to leave it as-is.

Reviewer 2: SGTM. Something that might help here is if we implement `ipfs key list -l {keyID}` since that would let us get the CID representation of the self key easily for publishing i.e. `ipfs key list -l self`.

Author: Good idea. I'll update this section with examples from https://github.com/ipfs/blog/pull/478

```

> :::

Reviewer: ```suggestion
```
